x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DefaultRoleReferenceResolver.java

```diff
@@ -0,0 +1,252 @@
+    }
+
+    @Override
+    public void resolveBwcApiKeyRoleReference(
+        RoleReference.BwcApiKeyRoleReference bwcApiKeyRoleReference,
+        ActionListener<RolesRetrievalResult> listener
+    ) {
+        final List<RoleDescriptor> roleDescriptors = apiKeyService.parseRoleDescriptors(
+            bwcApiKeyRoleReference.getApiKeyId(),
+            bwcApiKeyRoleReference.getRoleDescriptorsMap()
+        );
+        final RolesRetrievalResult rolesRetrievalResult = new RolesRetrievalResult();
+        rolesRetrievalResult.addDescriptors(Set.copyOf(roleDescriptors));
+        listener.onResponse(rolesRetrievalResult);
+    }
+
+    @Override
+    public void resolveServiceAccountRoleReference(
+        RoleReference.ServiceAccountRoleReference roleReference,
+        ActionListener<RolesRetrievalResult> listener
+    ) {
+        serviceAccountService.getRoleDescriptorForPrincipal(roleReference.getPrincipal(), listener.map(roleDescriptor -> {
+            final RolesRetrievalResult rolesRetrievalResult = new RolesRetrievalResult();
+            rolesRetrievalResult.addDescriptors(Set.of(roleDescriptor));
+            return rolesRetrievalResult;
+        }));
+    }
+
+    private void resolveRoleNames(Set<String> roleNames, ActionListener<RolesRetrievalResult> listener) {
+        roleDescriptors(roleNames, ActionListener.wrap(rolesRetrievalResult -> {
+            logDeprecatedRoles(rolesRetrievalResult.getRoleDescriptors());
+            final boolean missingRoles = rolesRetrievalResult.getMissingRoles().isEmpty() == false;
+            if (missingRoles) {
+                logger.debug(() -> new ParameterizedMessage("Could not find roles with names {}", rolesRetrievalResult.getMissingRoles()));
+            }
+            final Set<RoleDescriptor> effectiveDescriptors;
+            Set<RoleDescriptor> roleDescriptors = rolesRetrievalResult.getRoleDescriptors();
+            if (roleDescriptors.stream().anyMatch(RoleDescriptor::isUsingDocumentOrFieldLevelSecurity)
+                && DOCUMENT_LEVEL_SECURITY_FEATURE.checkWithoutTracking(licenseState) == false) {
+                effectiveDescriptors = roleDescriptors.stream()
+                    .filter(not(RoleDescriptor::isUsingDocumentOrFieldLevelSecurity))
+                    .collect(Collectors.toSet());
+            } else {
+                effectiveDescriptors = roleDescriptors;
+            }
+            logger.trace(
+                () -> new ParameterizedMessage(
+                    "Exposing effective role descriptors [{}] for role names [{}]",
+                    effectiveDescriptors,
+                    roleNames
+                )
+            );
+            effectiveRoleDescriptorsConsumer.accept(Collections.unmodifiableCollection(effectiveDescriptors));
+            logger.trace(
+                () -> new ParameterizedMessage("Building role from descriptors [{}] for role names [{}]", effectiveDescriptors, roleNames)
```

Comment:
It looks like we have 2 trace messages with essentially the same content here (with line 153 above)

Reply:
Right. Not sure why we had the 2nd one. I removed it.